src/widget/wsearchrelatedtracksmenu.cpp

```diff
@@ -0,0 +1,291 @@
+#include "widget/wsearchrelatedtracksmenu.h"
+
+#include "track/track.h"
+#include "util/qt.h"
+
+namespace {
+
+constexpr double kRelativeBpmRange = 0.06; // +/-6 %
+
+inline int bpmLowerBound(double bpm) {
+    return static_cast<int>(std::floor((1 - kRelativeBpmRange) * bpm));
+}
+
+inline int bpmUpperBound(double bpm) {
+    return static_cast<int>(std::ceil((1 + kRelativeBpmRange) * bpm));
+}
+
+QString extractCalendarYearNumberFromReleaseDate(
+        const QString& releaseDate) {
+    // TODO: Improve this poor calendar year number parser
+    // and extract the code
+    int skippedLeadingZeros = 0;
+    int countedLeadingDigits = 0;
+    const auto trimmed = releaseDate.trimmed();
+    // Count leading digits
+    for (int i = 0; i < trimmed.size(); ++i) {
+        if (!trimmed[i].isDigit()) {
+            break;
+        }
+        // Skip leading zeros
+        if (countedLeadingDigits == 0 && trimmed[i] == QChar('0')) {
+            ++skippedLeadingZeros;
+            continue;
+        }
+        // Count leading digits
+        ++countedLeadingDigits;
+    }
+    // Interpret the leading digits as the calendar year
+    // without any further validation
+    return trimmed.mid(skippedLeadingZeros, countedLeadingDigits);
+}
+
+} // namespace
+
+WSearchRelatedTracksMenu::WSearchRelatedTracksMenu(
+        QWidget* parent)
+        : QMenu(tr("Search related Tracks"), parent) {
+}
+
+bool WSearchRelatedTracksMenu::addTriggerSearchAction(
```

Comment:
```suggestion
void WSearchRelatedTracksMenu::addTriggerSearchAction(
```
Why does this return anything? I cannot find a reason. The return value is never used for anything.

Reply:
It is used if you look closely. This allows to write the invocation and resetting of the `addSeparatorBeforeNextAction` in a single line of code.

Reply:
I will pass it as a pointer instead (in/out param)

Reply:
Oh, I see that now. There's likely a more straightforward way to write this, but it's not immediately jumping out to me so I'm fine leaving it how it is.

Reply:
> I will pass it as a pointer instead (in/out param)

That seems like a better way.